fairscale/experimental/nn/multiprocess_pipe.py

```diff
@@ -3,204 +3,546 @@
+                    self.stream.wait_event(e)
+
+        # Determine whether checkpointing or not.
+        checkpoint = chunk < checkpoint_stop
+        if checkpoint:
+
+            def function(input: TensorOrTensors, chunk_id: int = chunk) -> TensorOrTensors:
+                with record_function("chunk%d-rank%d" % (chunk_id, dist_record.rank)):
+                    result = self.module(*input)
+                    if self.num_output is None:
+                        result = (result,)
+                    return tuple(result)
+
+            chk = Checkpointing(function, batch)
+            task = Task(self.stream, compute=chk.checkpoint, finalize=chk.recompute)
+            del function, chk
+
+        else:
+
+            def compute(
+                batch: Batch = batch,
+                chunk_id: int = chunk,
+                rank: int = dist_record.rank if dist_record is not None else -1,
+            ) -> Batch:
+                with record_function("chunk%d-rank%d" % (chunk_id, dist_record.rank)):
+                    result = self.module(*batch.tensors)
+                    if self.num_output is None:
+                        result = (result,)
+                return Batch(result, chunk_id)
+
+            task = Task(self.stream, compute=compute, finalize=None)
+            del compute
+
+        self.in_queue.put(task)
+
+        ok, payload = self.out_queue.get()
+
+        # Hold the first exception.
+        if exc_info is not None:
+            pass
+        elif not ok:
+            exc_info = cast(ExcInfo, payload)
+        else:
+            task, batch = cast(Tuple[Task, Batch], payload)
+
+            with use_device(self.device):
+                task.finalize(batch)
+
+            dist_record.batches[chunk] = batch
+
+        if exc_info is not None:
+            raise exc_info[0].with_traceback(exc_info[1], exc_info[2])
+
+    def forward_results(self, chunk: int, dist_record: DistributedPipelineRecord) -> None:
+        with use_stream(self.stream):
+            for user, input_idx, output_idx in dist_record.users:
+                v = dist_record.get_batch(chunk).value[output_idx]
+                dist_record.forwarded_phony[chunk][output_idx].append(user.remote().feed(chunk, input_idx, v))
+
+    def run_pipeline(self, dist_record_ref: rpc.RRef) -> Optional[Tensor]:
+        dist_record = dist_record_ref.local_value()
+        self.run(dist_record)
+
+        if not dist_record.users:
+            result = microbatch.gather(dist_record.batches)
+            assert len(result) == 1
+            result = result[0]
+            s0 = current_stream(result.device)
+            if is_cuda(s0):
+                # TODO. Investigate why this is needed and remove it if possible.
+                as_cuda(s0).synchronize()
+            return result
+
+        return None
+
+
+class MultiInputSequential(nn.Module):
+    """
+        A variation of nn.Sequential, that allows the first module in the sequence accepts
```

Comment:
Why only the first module?

Reply:
Please note, this class is only for internal use in this file. Currently the _split_module function packs consecutive layers into a single partition if they are on the same device, and some conditions hold, including that only the first of such layers may be a multi-input layer , and only last of them can be multi-output (i.e. it breaks the partition each time it seems a multi-input or multi-output layer).  Thus, in current implementation only the first layer could be multi-input.